rasa/core/processor.py

```diff
@@ -374,8 +376,23 @@ def predict_next_action(
         """Predicts the next action the bot should take after seeing x.
 
         This should be overwritten by more advanced policies to use
-        ML to predict the action. Returns the index of the next action.
+        ML to predict the action.
+
+        Returns:
+             The index of the next action and prediction of the policy.
+
+        Raises:
+            ActionLimitReached if the limit of actions to predict has been reached.
         """
+        should_predict_another_action = self.should_predict_another_action(
+            tracker.latest_action_name
+        )
+
+        if self.is_action_limit_reached(tracker, should_predict_another_action):
```

Comment:
I still think this is confusing as we raise `ActionLimitReached` even if it's not been reached and `should_predict_another_action` is just `False`. I think the only reason this doesn't cause problems later is because we _also_ check `self.should_predict_another_action` in `_run_action`. Otherwise we would be tripping `on_circuit_break` every time we got `action_listen`!

But honestly looking through all the code, it will be quite tricky to get all of this to work cleanly - and as we said before, all of this processor implementation will go!
I think a good compromise would be to rename  `ActionLimitReached` to something more generic such as `StopPredictionActions` and then put a comment above `if self.on_circuit_break:` stating that this will only be raised in this loop if action limit is reached.

How does that sound?

Reply:
Sorry disregard this - I miss-read the implementation of `is_action_limit_reached`